### PR TITLE
[9.0][stock_operating_unit]

### PR DESCRIPTION
**Pull request by JordiBForgeFlow**
Supersedes https://github.com/OCA/operating-unit/pull/20

### CODE REVIEW

**Comment by JordiBForgeFlow** [comment]
cc @aheficent Can you review?

**Comment by AaronHForgeFlow** [comment]
IMHO this is ready to be merged :+1:

File: stock_operating_unit/model/stock.py
Comment (angelmoya) [inline stock_operating_unit/model/stock.py]: you could use:
 '=', rec.id
instead of:
'in', rec.ids
rec is one, so you do not need to use "in"

File: stock_operating_unit/model/stock.py
Comment (angelmoya) [inline stock_operating_unit/model/stock.py]: you could use one for w in warehouse and test tree conditions instead create tree for for each condition.

File: stock_operating_unit/model/stock.py
Comment (angelmoya) [inline stock_operating_unit/model/stock.py]: It would be nice to use more clean names.
sm -> move or stock_move

**Review by angelmoya — COMMENTED** [review]
Code review done, I will test on runbot now.

Some remarks but it's ok

File: stock_operating_unit/model/stock.py
Comment (angelmoya) [inline stock_operating_unit/model/stock.py]: Why do not put this constraint on stock moves?

**Comment by angelmoya** [comment]
On stock picking lines, I think that:
- Source Location Operating Unit 
- Dest. Location Operating Unit 

are to long labels for user interface

File: stock_operating_unit/model/stock.py
Comment (AaronHForgeFlow) [reply stock_operating_unit/model/stock.py]: Yes, but as rec.ids is a list i got to use rec.ids[0]

File: stock_operating_unit/model/stock.py
Comment (AaronHForgeFlow) [reply stock_operating_unit/model/stock.py]: All right. Thank you Ángel

File: stock_operating_unit/model/stock.py
Comment (AaronHForgeFlow) [reply stock_operating_unit/model/stock.py]: Done

**Comment by AaronHForgeFlow** [comment]
About the long names I think that there are long, however if we shorten them it will be confusing.

**Comment by manuelfcordoba** [comment]
Creating a delivery order I noticed that the Requesting Operating Unit by default is set to the Main Operating Unit. I think it should be =false by default.

Creating a delivery order throw the kan ban view of an specific Picking type and changing the Picking type in the form view generated an inconsintence on the Picking type of the moves.

**Comment by AaronHForgeFlow** [comment]
Travis Lint is failing because the onchange_picking_type method is in v7 in the odoo core, so it has to be overwritten this way.

**Review by LoisRForgeFlow — COMMENTED** [review]
When accessing to the stock locations as stock manager with Demo User and the B2B Operating Unit assigned, the next access error pops:
image
I think the stock manager of a OU should have access to the locations of his/her warehouse at least.

**Comment by AaronHForgeFlow** [comment]
I also realize that when creating stock pickings, if the user changes the destination location the system does not change the requesting operating unit and the stock picking

**Comment by AaronHForgeFlow** [comment]
Regarding my last comment. The Picking type is the one that controls the operating unit of the picking. If the user changes the destination location the system automatically changes the destination location operating unit, so it was already correct.

**Review by AaronHForgeFlow — APPROVED** [review]
LGTM now :+1:

**Review by LoisRForgeFlow — APPROVED** [review]
Tested on runbot + code review :+1:

**Comment by AaronHForgeFlow** [comment]
@angelmoya can you review?

File: stock_operating_unit/__openerp__.py
Comment (guewen) [inline stock_operating_unit/__openerp__.py]: The following form would be better:

```
    "summary": "An operating unit (OU) is an organizational entity part of a "
               "company",
```

Because ithe current form includes a lot of spaces before "company".

File: stock_operating_unit/demo/stock_demo.xml
Comment (guewen) [inline stock_operating_unit/demo/stock_demo.xml]: `<openerp><data noupdate="1">` can be replaced by `<odoo noupdate="1">`

File: stock_operating_unit/model/stock.py
Comment (guewen) [inline stock_operating_unit/model/stock.py]: `_` can be imported from the same package: `from openerp import _, api, fields, models`

File: stock_operating_unit/model/stock.py
Comment (guewen) [inline stock_operating_unit/model/stock.py]: Better: `from openerp.exceptions import UserError`. `Warning` is deprecated.

File: stock_operating_unit/model/stock.py
Comment (guewen) [inline stock_operating_unit/model/stock.py]: `self.env.uid` is a cleaner (no access to a "private" attribute).

You would have less line breaks with such indentation:

```
    operating_unit_id = fields.Many2one(
        comodel_name='operating.unit',
        string='Operating Unit',
        default=lambda self: self.env['res.users'].
                operating_unit_default_get(self.env.uid))
    )
```

File: stock_operating_unit/model/stock.py
Comment (guewen) [inline stock_operating_unit/model/stock.py]: `\` should be avoided when it parenthesis can be used:

```python

if (rec.company_id and rec.operating_unit_id and 
    rec.company_id != rec.operating_unit_id.company_id):



```

File: stock_operating_unit/model/stock.py
Comment (guewen) [inline stock_operating_unit/model/stock.py]: This string contains a lot of spaces which is then displayed weirdly in the error message. You should use the string concatenation:

```python

raise UserError(
    _('Configuration error\nThe Company in the '
      'Stock Warehouse and in the Operating Unit must be the same.')
)

```

(note: I also removed the exclamation point in the message which is not recommended (https://cybertext.wordpress.com/2011/11/18/error-messages-that-make-you-feel-stupid/)

**Review by guewen — COMMENTED** [review]
I started to leave comments but I'll submit a PR to make the changes directly.

**Comment by AaronHForgeFlow** [comment]
@guewen should I close this?

**Comment by guewen** [comment]
@aheficent nope, here is my proposal that you can merge in yours if you agree https://github.com/Eficent/operating-unit/pull/64

**Comment by AaronHForgeFlow** [comment]
All right, Thank you @guewen

**Review by guewen — APPROVED** [review]
Thanks!
There are enough approved reviews to merge.

Note: I have the rights to merge but I don't want to walk over the Operating Units PSC

**Comment by guewen** [comment]
Merging, it will unblock `procurement_operating_unit` (which in turns will unblock `purchase_operating_unit` once merged).

**Pull request merged**
